stlab/channel.hpp

```diff
@@ -464,13 +537,141 @@ struct shared_process : shared_process_receiver<yield_type<T, Arg>>,
 
 /**************************************************************************************************/
 
-template <typename T>
-auto channel() -> std::pair<sender<T>, receiver<T>> {
-    auto p = std::make_shared<detail::shared_process<identity, T>>(identity());
+template <typename T, typename S>
+auto channel(S&& s) -> std::pair<sender<T>, receiver<T>> {
+    auto p = std::make_shared<detail::shared_process<identity, T>>(std::forward<S>(s), identity());
     return std::make_pair(sender<T>(p), receiver<T>(p));
 }
 
 /**************************************************************************************************/
+namespace detail {
+
+template <typename... T>
+struct join_context : std::enable_shared_from_this<join_context<T...>>
+{
+    std::tuple<T...>    _result;
+    std::atomic_size_t  _remaining{ sizeof...(T) };
+    stlab::process_state _state = process_state::await;
+
+
+    template<std::size_t i, typename U>
+    void await(U&& u) {
+        std::get<i>(_result) = std::forward<U>(u);
```

Comment:
What ensures that for every _i_ this gets called only once before yielding a value?

r = join([](int x, int y){ cout << x << y << endl; }, r1, r2);

At the beginning, from the perspective of r1 and r2, the join is in an await state. If it receives a value from r1, then r1 needs to the receiving process in what amounts to a still running state even though it isn't running but rather is awaiting a value from r2. I think this requires changes at the shared_process level.